Add possibility to override node package at cluster creation and update

jdeamicis opened the pull request:
### Description of changes
* Add possibility to override node package at cluster creation and update
* Move installation of custom node package to ad-hoc recipe

### Tests
* Manually created a cluster with custom node package
* Manually updated a cluster with custom node package

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- ~[ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.~
- ~[ ] Check if documentation is impacted by this change.~

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Add possibility to override node package at cluster creation and update

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

lukeseawalker commented on cookbooks/aws-parallelcluster-install/recipes/base.rb:
```diff
@@ -146,27 +146,7 @@
-      deactivate
-    NODE
-  end
+  include_recipe 'aws-parallelcluster-install::parallelcluster_custom_node'
```
I'd remove `_custom_`, we are installing `parallelcluster-node`

lukeseawalker replied:
nevermind, I see that only the portion to install custom node was moved.
Can we move everything in the new cookbook so we have only one place where the node is installed? This would be cleaner and more maintainable

lukeseawalker commented on cookbooks/aws-parallelcluster-config/recipes/update.rb:
```diff
@@ -24,3 +24,18 @@
+    action :stop
+  end
+
+  include_recipe 'aws-parallelcluster-install::parallelcluster_custom_node'
```
isn't better to postpone the node installation before the slurm::update? This because during the slurm update the clustermgtd will be restarted

jdeamicis replied:
That happens only if the queues are updated, right? This PR instead is for testing the node logic, which is independent of the cluster configuration.

lukeseawalker commented on cookbooks/aws-parallelcluster-config/recipes/update.rb:
```diff
@@ -24,3 +24,18 @@
+if !node['cluster']['custom_node_package'].nil? && !node['cluster']['custom_node_package'].empty?
+
+  service "supervisord" do
+    action :stop
```
why stopping the whole supervisord and not just the demon provided by the node package.
Furthermore, are you sure that you are targeting the right supervisord? AFAIK it should be the one present in the cookbook virtualenv, not the one installed at OS level

jdeamicis replied:
> why stopping the whole supervisord and not just the demon provided by the node package.

Doing it.

> Furthermore, are you sure that you are targeting the right supervisord? AFAIK it should be the one present in the cookbook virtualenv, not the one installed at OS level

By switching to restarting only clustermgtd, this point becomes obsolete.

jdeamicis replied:
Done.

jdeamicis committed:
Fix inconsistency in recipes after rebasing.

A node.rb recipe was recently created, so the logic for the custom
node installation must be removed from the base.rb recipe.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Rename parallelcluster_custom_node.rb to custom_parallelcluster_node.rb

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Rename node.rb to parallelcluster_node.rb

Add TODO in custom_parallelcluster_node.rb

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Stop and restart only clustermgtd during a custom node package upgrade

Rather than stopping the whole supervisord daemon, only stop and restart
the clustermgtd service managed by supervisord.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
